presto-orc/src/main/java/com/facebook/presto/orc/StreamLayout.java

```diff
@@ -0,0 +1,101 @@
+
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+import static java.util.Objects.requireNonNull;
+
+/**
+ * StreamLayout is used by the Writer to reorder the data streams.
+ * In ORC, a column is stored in multiple streams (for example
+ * DATA, PRESENT, LENGTH, DICTIONARY and so on).
+ */
+public interface StreamLayout
+{
+    void reOrder(List<StreamDataOutput> dataStreams);
+
+    /**
+     * Streams are ordered by the Stream Size. There is no ordering between
+     * two streams of the same size. This strategy places all small streams
+     * together, so that multiple of these small IOs can be combined. Note:
+     * usually columns contain small (e.g. PRESENT) and large streams (e.g. DATA).
+     * This strategy places them far apart and may result in increased IO.
+     */
+    class ByStreamSize
+            implements StreamLayout
+    {
+        @Override
+        public void reOrder(List<StreamDataOutput> dataStreams)
+        {
+            requireNonNull(dataStreams, "dataStreams is null");
+            Collections.sort(dataStreams);
+        }
+    }
+
+    /**
+     * Streams are ordered by Column Size. If two columns have same size then
+     * columnId, stream size and stream kind are used for ordering. This strategy
+     * places small columns together to read multiple small columns in one IO.
+     * All streams for a column are stored together to read one column in one IO.
+     */
+    class ByColumnSize
+            implements StreamLayout
+    {
+        @Override
+        public void reOrder(List<StreamDataOutput> dataStreams)
+        {
+            requireNonNull(dataStreams, "dataStreams is null");
+            if (dataStreams.isEmpty()) {
+                return;
+            }
+
+            final Map<Integer, Long> columnSizes = dataStreams.stream()
+                    .collect(Collectors.toMap(
+                            s -> s.getStream().getColumn(),
+                            s -> (long) s.getStream().getLength(),
+                            (s1, s2) -> s1 + s2));
+
+            dataStreams.sort((sd1, sd2) -> {
+                Stream s1 = sd1.getStream();
+                Stream s2 = sd2.getStream();
+
+                long sizeDelta = columnSizes.get(s1.getColumn()) -
+                        columnSizes.get(s2.getColumn());
+                if (sizeDelta != 0) {
+                    return sizeDelta < 0 ? -1 : 1;
+                }
+
+                // order by column attributes
+                int columnDelta = s1.getColumn() - s2.getColumn();
+                if (columnDelta != 0) {
+                    return columnDelta;
+                }
+
+                sizeDelta = s1.getLength() - s2.getLength();
+                if (sizeDelta != 0) {
+                    return sizeDelta < 0 ? -1 : 1;
+                }
+
+                return s1.getStreamKind().compareTo(s2.getStreamKind());
```

Comment:
Just a thought: I wonder how stream kind and stream length interact with each other in the reader, and does it really make sense to put the kind last.

Reply:
Stream length can be predicted to a certain extent with Stream kind. Present streams/in dictionary streams have less data when compared to data streams for most cases. Stream Kind is the last tie breaker to get predictable order of the column size.